### PR TITLE
🐛 fix inmemory provider docker build

### PR DESCRIPTION
**Pull request by ykakarap**
**What this PR does / why we need it**:

Fixes the bug in the current make target that builds the in-memory infrastructure provider.
The current bug: It builds the wrong image. It rebuild the cluster-api-controller-manager image.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: test/infrastructure/inmemory/Dockerfile
Comment (fabriziopandini) [inline test/infrastructure/inmemory/Dockerfile:1]: This File has been dropped in a recent PR, we are using the docker file from the root folder

Comment (sbueringer) [reply test/infrastructure/inmemory/Dockerfile:1]: Yup this PR re-introduces the file. I think because Yuvaraj thought that it's needed for the fix. I think it's not > https://github.com/kubernetes-sigs/cluster-api/pull/8822/files#r1223985730

File: test/infrastructure/inmemory/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+COPY . .
+
+# Essentially, change directories into CAPIM
+WORKDIR /workspace/test/infrastructure/inmemory
+
+# Build
+ARG ARCH
+
+# Build the CAPIM manager using the compiler cache folder
+RUN --mount=type=cache,target=/root/.cache/go-build \
```
Comment (sbueringer) [inline test/infrastructure/inmemory/Dockerfile:64]: Let's please reduce the diff to the top-level Dockerfile as far as possible, e.g.:
* support lflags arg
* use nonroot image
* probably more

File: test/infrastructure/inmemory/Dockerfile
Comment (sbueringer) [reply test/infrastructure/inmemory/Dockerfile:1]: Realized that it's not as simple as re-using one of the other Dockerfiles: https://github.com/kubernetes-sigs/cluster-api/pull/8822#discussion_r1226673882

File: test/infrastructure/inmemory/Dockerfile (hunk 1 above)
Comment (sbueringer) [reply test/infrastructure/inmemory/Dockerfile:64]: WDTY would it be possible to use the same Dockerfile but have a some args for the differences? (can be a follow-up issue, we need to fix the current issue ASAP in my opinion)

Comment (ykakarap) [reply test/infrastructure/inmemory/Dockerfile:68]: I changed the Dockerfile to be as close to the root level Dockerfile as possible.

File: test/infrastructure/inmemory/Dockerfile
Comment (ykakarap) [reply test/infrastructure/inmemory/Dockerfile:1]: I did some digging and it looks like the cleanest solution is to have a separate dockerfile and not try to reuse the root level Dockerfile. Since the in-memory provider is part of the test module which has it owns go module it is different from CABPK, KCP, etc and therefore needs to be treated a differently. Hence I would suggest not trying to re-use the root level package.


An option we can explore is having a single Dockerfile for the test package and reusing that for all the test providers: Docker, in-memory. But since the docker provider needs root privilege and is built a little differently I am not sure how useful that might be. I would not want this to block the current PR. This might be something to explore in a separate issue and not here.

TLDR; The cleaner solution seems to be just have its own Dockerfile and not try to reuse the root level Dockerfile.

Comment (sbueringer) [reply test/infrastructure/inmemory/Dockerfile:1]: Thx for taking a closer look. I missed a few of those details at a first glance. Fine for me to just use a separate Dockerfile for now.
